cloudify_cli/bootstrap/bootstrap.py

```diff
@@ -322,3 +333,124 @@ def _copy_agent_key(agent_local_key_path, agent_remote_key_path,
+
+            if not plugin_local_yaml_path or not plugin_remote_yaml_path:
+                missing_fields = []
+                if plugin_local_yaml_path is None:
+                    missing_fields.append('source_path')
+                if plugin_remote_yaml_path is None:
+                    missing_fields.append('destination_path')
+
+                raise \
+                    CloudifyBootstrapError('The following fields are missing: '
+                                           '{0}.'
+                                           .format(','.join(missing_fields)))
+
+            if plugin_remote_yaml_path.startswith('/'):
+                plugin_remote_yaml_path = plugin_remote_yaml_path[1:]
+
+            plugin_local_yaml_path = \
+                _get_resource_into_dir(temp_dir, plugin_local_yaml_path)
+
+            # copy plugin's yaml file to the manager's fileserver
+            remote_paths = plugin_remote_yaml_path.split('/')
+            with fabric.settings(**fabric_env):
+
+                # Creating a remote folder
+                if len(remote_paths) > 1:
+                    remote_plugin_folder_path = \
+                        os.path.join(remote_plugins_folder, *remote_paths[:-1])
+                    fabric.run('sudo mkdir -p {0}'.format(
+                        remote_plugin_folder_path))
+
+                # Uploading the plugin file
+                remote_plugin_path = os.path.join(remote_plugins_folder,
+                                                  *remote_paths)
+                fabric.put(plugin_local_yaml_path,
+                           remote_plugin_path, use_sudo=True)
+    finally:
+        shutil.rmtree(temp_dir)
+
+
+def _get_resource_into_dir(destination_dir, resource_source_path):
+    """
+    Copies a given path into the destination dir. The path could refer to
+    a local file or a remote url. NOTE: If two files shares the same name,
+    the old file would be overwritten.
+    :param destination_dir: The dir to write the resource into.
+    :param resource_source_path: A path to the resource.
+    :return: the path to the newly copied resource.
+    """
+    logger = get_logger()
+    temp_file_path = os.path.join(destination_dir,
+                                  resource_source_path.split('/')[-1])
+
+    parts = urlparse.urlsplit(resource_source_path)
+
+    if not parts.scheme or not parts.netloc:
+        if not os.path.isabs(resource_source_path):
+            resource_source_path = os.path.abspath(resource_source_path)
+        logger.info('Copying from {0} to {1}'
+                    .format(resource_source_path, temp_file_path))
+        shutil.copyfile(resource_source_path, temp_file_path)
+        logger.info('Done copying')
+    else:
+        try:
+            logger.info('Downloading from {0} to {1}'
+                        .format(resource_source_path, destination_dir))
+            remote_resource = requests.get(resource_source_path)
```

Comment:
http://stackoverflow.com/questions/13137817/how-to-download-image-using-requests